integrations/firehydrant/client.py

```diff
@@ -0,0 +1,451 @@
+                f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+            )
+            raise
+
+    async def send_paginated_api_request(
+        self,
+        endpoint: str,
+        data_key: str,
+        method: str = "GET",
+        query_params: Optional[dict[str, Any]] = None,
+        json_data: Optional[dict[str, Any]] = None,
+    ) -> list[dict[str, Any]]:
+        """
+        Sends an API request to Firehydrant and returns all resources available. It does not use generator
+
+        :param endpoint: API endpoint URL
+        :param data_key: Resource key to fetch
+        :param method: HTTP method (default: 'GET')
+        :param query_params: Query parameters (default: None)
+        :param json_data: JSON data to send in request body (default: None)
+        :return: Response JSON data
+        """
+        try:
+            all_resources = []  # List to hold all fetched resources
+
+            while True:
+                response = await self.http_client.request(
+                    method=method,
+                    url=f"{self.base_url}/v1/{endpoint}",
+                    params=query_params,
+                    json=json_data,
+                    headers=self.api_auth_header,
+                )
+                response.raise_for_status()
+                response_json = response.json()
+
+                all_resources.extend(response_json.get(data_key, []))
+
+                # Check for paging information and decide whether to fetch more pages
+                paging_info = response_json.get("pagination")
+                if paging_info and paging_info.get("page", 0) < paging_info.get(
+                    "pages", 0
+                ):
+                    query_params = query_params or {}
+                    query_params["page"] = paging_info["page"] + 1
+                else:
+                    break
+
+            return all_resources
+
+        except httpx.HTTPStatusError as e:
+            print(
+                f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+            )
+            raise
+
+    async def get_paginated_resource(
+        self, resource_type: str
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        """
+        Fetch paginated resources of a specific type from the Firehydrant API.
+
+        Args:
+            resource_type (str): The type of resource to fetch, e.g., "incidents" or "webhooks."
+
+        Yields:
+            AsyncGenerator[list[dict[str, Any]], None]: A generator that yields lists of resource data.
+                Each yielded list contains a batch of resources of the specified type.
+
+        Returns:
+            None: This function does not return a value but yields resource data via asynchronous iteration.
+        """
+        logger.info(f"Getting {resource_type} from Firehydrant")
+        endpoint = endpoint_resource_type_mapper.get(resource_type, "")
```

Comment:
is it possible for resource type to not be one of the known types?

the request will fail for that url`

Reply:
The resource types are all the known firehydrant types we are synching: environments, services, incidents and retrospectives